File: src/main/java/com/library/models/TransactionEmprunt.java
package com.library.models;

import java.time.LocalDate;
import java.time.temporal.ChronoUnit;

/**
 * Classe TransactionEmprunt
 * Enregistre une transaction d'emprunt avec ses dates et statut
 */
public class TransactionEmprunt {
    private String idTransaction;
    private String idUtilisateur;
    private String isbnLivre;
    private LocalDate dateEmprunt;
    private LocalDate dateRetourPrevue;
    private LocalDate dateRetourEffective;
    private static final long DUREE_EMPRUNT_JOURS = 14;

    public TransactionEmprunt(String idTransaction, String idUtilisateur, String isbnLivre) {
        this.idTransaction = idTransaction;
        this.idUtilisateur = idUtilisateur;
        this.isbnLivre = isbnLivre;
        this.dateEmprunt = LocalDate.now();
        this.dateRetourPrevue = dateEmprunt.plus(DUREE_EMPRUNT_JOURS, ChronoUnit.DAYS);
        this.dateRetourEffective = null;
    }

    public String getIdTransaction() {
        return idTransaction;
    }

    public String getIdUtilisateur() {
        return idUtilisateur;
    }

    public String getIsbnLivre() {
        return isbnLivre;
    }

    public LocalDate getDateEmprunt() {
        return dateEmprunt;
    }

    public LocalDate getDateRetourPrevue() {
        return dateRetourPrevue;
    }

    public LocalDate getDateRetourEffective() {
        return dateRetourEffective;
    }

    public void setDateRetourEffective(LocalDate date) {
        this.dateRetourEffective = date;
    }

    public boolean isEnRetard() {
        if (dateRetourEffective != null) {
            return dateRetourEffective.isAfter(dateRetourPrevue);
        }
        return LocalDate.now().isAfter(dateRetourPrevue);
    }

    public long getJoursRestants() {
        return ChronoUnit.DAYS.between(LocalDate.now(), dateRetourPrevue);
    }

    public long getJoursDeRetard() {
        if (isEnRetard()) {
            LocalDate dateComparaison = dateRetourEffective != null ? 
                    dateRetourEffective : LocalDate.now();
            return ChronoUnit.DAYS.between(dateRetourPrevue, dateComparaison);
        }
        return 0;
    }

    public boolean isRetourne() {
        return dateRetourEffective != null;
    }

    @Override
    public String toString() {
        return "TransactionEmprunt{" +
                "idTransaction='" + idTransaction + '\'' +
                ", idUtilisateur='" + idUtilisateur + '\'' +
                ", isbnLivre='" + isbnLivre + '\'' +
                ", dateEmprunt=" + dateEmprunt +
                ", dateRetourPrevue=" + dateRetourPrevue +
                (dateRetourEffective != null ? ", dateRetourEffective=" + dateRetourEffective : "") +
                ", enRetard=" + (isEnRetard() ? "OUI" : "NON") +
                '}';
    }
}

File: js/src/models/TransactionEmprunt.js
/**
 * Classe TransactionEmprunt
 * Enregistre une transaction d'emprunt avec ses dates et statut
 */
class TransactionEmprunt {
    constructor(idTransaction, idUtilisateur, isbnLivre) {
        this.idTransaction = idTransaction;
        this.idUtilisateur = idUtilisateur;
        this.isbnLivre = isbnLivre;
        this.dateEmprunt = new Date();
        
        // Date de retour prévue: 14 jours après l'emprunt
        this.dateRetourPrevue = new Date();
        this.dateRetourPrevue.setDate(this.dateRetourPrevue.getDate() + 14);
        
        this.dateRetourEffective = null;
    }

    static DUREE_EMPRUNT_JOURS = 14;

    getIdTransaction() {
        return this.idTransaction;
    }

    getIdUtilisateur() {
        return this.idUtilisateur;
    }

    getIsbnLivre() {
        return this.isbnLivre;
    }

    getDateEmprunt() {
        return this.dateEmprunt;
    }

    getDateRetourPrevue() {
        return this.dateRetourPrevue;
    }

    getDateRetourEffective() {
        return this.dateRetourEffective;
    }

    setDateRetourEffective(date) {
        this.dateRetourEffective = date;
    }

    /**
     * Vérifier si le livre est en retard
     * @returns {boolean} True si en retard
     */
    isEnRetard() {
        const dateComparaison = this.dateRetourEffective || new Date();
        return dateComparaison > this.dateRetourPrevue;
    }

    /**
     * Obtenir le nombre de jours restants
     * @returns {number} Jours restants (peut être négatif si en retard)
     */
    getJoursRestants() {
        const aujourd = new Date();
        const diffTime = this.dateRetourPrevue - aujourd;
        return Math.ceil(diffTime / (1000 * 60 * 60 * 24));
    }

    /**
     * Obtenir le nombre de jours de retard
     * @returns {number} Jours de retard (0 si pas de retard)
     */
    getJoursDeRetard() {
        if (this.isEnRetard()) {
            const dateComparaison = this.dateRetourEffective || new Date();
            const diffTime = dateComparaison - this.dateRetourPrevue;
            return Math.ceil(diffTime / (1000 * 60 * 60 * 24));
        }
        return 0;
    }

    /**
     * Vérifier si le livre a été retourné
     * @returns {boolean} True si retourné
     */
    isRetourne() {
        return this.dateRetourEffective !== null;
    }

    /**
     * Formater une date
     * @param {Date} date - La date à formater
     * @returns {string} Date formatée (YYYY-MM-DD)
     */
    static formatDate(date) {
        if (!date) return null;
        return date.toISOString().split('T')[0];
    }

    toString() {
        let str = `TransactionEmprunt{
            idTransaction: '${this.idTransaction}',
            idUtilisateur: '${this.idUtilisateur}',
            isbnLivre: '${this.isbnLivre}',
            dateEmprunt: ${TransactionEmprunt.formatDate(this.dateEmprunt)},
            dateRetourPrevue: ${TransactionEmprunt.formatDate(this.dateRetourPrevue)}`;
        
        if (this.dateRetourEffective) {
            str += `,\n            dateRetourEffective: ${TransactionEmprunt.formatDate(this.dateRetourEffective)}`;
        }
        
        str += `,\n            enRetard: ${this.isEnRetard() ? 'OUI' : 'NON'}`;
        str += '\n        }';
        
        return str;
    }
}

module.exports = TransactionEmprunt;
